vault/identity_store_oidc_test.go

```diff
@@ -0,0 +1,425 @@
+package vault
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/go-test/deep"
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/sdk/logical"
+	"gopkg.in/square/go-jose.v2"
+)
+
+// TestOIDC_Path_OIDCRoleRole tests CRUD operations for roles
+func TestOIDC_Path_OIDCRoleRole(t *testing.T) {
+	c, _, _ := TestCoreUnsealed(t)
+	ctx := namespace.RootContext(nil)
+	storage := &logical.InmemStorage{}
+
+	// Create a test key "test-key"
+	c.identityStore.HandleRequest(ctx, &logical.Request{
+		Path:      "oidc/key/test-key",
+		Operation: logical.CreateOperation,
+		Storage:   storage,
+	})
+
+	// Create a test role "test-role1" with a valid key -- should succeed
+	resp, err := c.identityStore.HandleRequest(ctx, &logical.Request{
+		Path:      "oidc/role/test-role1",
+		Operation: logical.CreateOperation,
+		Data: map[string]interface{}{
+			"key": "test-key",
+		},
+		Storage: storage,
+	})
+	expectSuccess(t, resp, err)
+
+	// Read "test-role1"
+	respReadTestRole1, err1 := c.identityStore.HandleRequest(ctx, &logical.Request{
+		Path:      "oidc/role/test-role1",
+		Operation: logical.ReadOperation,
+		Storage:   storage,
+	})
+	expectSuccess(t, respReadTestRole1, err1)
+
+	// Create a test role "test-role2" witn an invalid key -- should fail
+	resp, err = c.identityStore.HandleRequest(ctx, &logical.Request{
+		Path:      "oidc/role/test-role2",
+		Operation: logical.CreateOperation,
+		Data: map[string]interface{}{
+			"key": "a-key-that-does-not-exist",
+		},
+		Storage: storage,
+	})
+	expectError(t, resp, err)
+
+	// Update "test-role1" with valid parameters -- should succeed
+	resp, err = c.identityStore.HandleRequest(ctx, &logical.Request{
+		Path:      "oidc/role/test-role1",
+		Operation: logical.UpdateOperation,
+		Data: map[string]interface{}{
+			"template": "{\"some-key\":\"some-value\"}",
+			"ttl":      "2h",
+		},
+		Storage: storage,
+	})
+	expectSuccess(t, resp, err)
+
+	// Read "test-role1" again
+	respReadTestRole1AfterUpdate, err2 := c.identityStore.HandleRequest(ctx, &logical.Request{
+		Path:      "oidc/role/test-role1",
+		Operation: logical.ReadOperation,
+		Storage:   storage,
+	})
+	expectSuccess(t, respReadTestRole1AfterUpdate, err2)
+
+	// Compare response for "test-role1" before and after it was updated
+	expectedDiff := map[string]interface{}{
```

Comment:
This approach seems pretty rigidly tied to `deep` (e.g. they tweak their presentation and the tests break), and I'm not sure I see the need to check what the value was. Isn't is simpler to just have an expected `map[string]interface{}` and `deep.Diff` that?